### PR TITLE
Fixed #31055 -- Handled database errors on check constraint support check.

### PR DESCRIPTION
**Pull request by charettes**
Since #28478 test databases are only created if necessary for the subset of
tests targeted.

Under such circumstances its possible for some databases defined in
settings.DATABASES not to exist which prevents the retrieval of the
supports_table_check_constraints feature flag on MySQL since it requires a
usable connection.

Skipping the check when this happens should not be harmful given it will run
under normal circumstances when the full set of tests are targeted or when
checks are run from a different management command.

---

https://code.djangoproject.com/ticket/31055

### CODE REVIEW

File: django/db/models/base.py [hunk 1]
```diff
@@ -1838,8 +1838,13 @@ def _check_constraints(cls):
...
                 continue
             connection = connections[db]
+            try:
+                supports_table_check_constraints = connection.features.supports_table_check_constraints
+            except DatabaseError:
+                # Could not establish a connection, disable check instead of crashing.
+                continue
             if (
-                connection.features.supports_table_check_constraints or
+                supports_table_check_constraints or
```
Comment (felixxm) [inline django/db/models/base.py]: I'm afraid that we can hit this in other cases, e.g. in [_check_json_fields()](https://github.com/django/django/pull/11452/files#diff-507b415116b409afa4f723e41a759a9eR1340-R1360) in #11452. What do you think about checking connection usability in `mysql_server_info()`, e.g.
```python
    @cached_property
    def mysql_server_info(self):
        if not self.is_usable():
            return ''
        with self.temporary_connection() as cursor:
            cursor.execute('SELECT VERSION()')
            return cursor.fetchone()[0]
```

Comment (charettes) [reply django/db/models/base.py]: @felixxm that's what I initially tried but it was crashing with `AttributeError`

```
  File "/tests/django/django/db/backends/mysql/base.py", line 351, in mysql_server_info
    if not self.is_usable():
  File "/tests/django/django/db/backends/mysql/base.py", line 329, in is_usable
    self.connection.ping()
AttributeError: 'NoneType' object has no attribute 'ping'
```

Comment (felixxm) [reply django/db/models/base.py]: So maybe we should check connection in `is_usable()` :thinking: like in related fix for PostgreSQL
https://github.com/django/django/commit/a96b9019320ed8236659ee520a7a017c1bafbc6f#diff-aef216ccdc37941bfed98ea3c80f7642R251-R252

**Comment by charettes** [comment]
@felixxm it looks like we don't run `database` tagged by default for this exact reason 

https://github.com/django/django/blame/8d087f9a179c287e5bb1e24558b097e0a761bc28/django/core/checks/registry.py#L67-L69

And we only run these tests on `migrate` which seems more appropriate

https://github.com/django/django/blob/8d087f9a179c287e5bb1e24558b097e0a761bc28/django/core/management/commands/migrate.py#L62-L65

The solution is likely to write these two tests the other way around by iterating over models using check constraints when the feature is not available. Thoughts?

File: django/db/models/constraints.py [hunk 2]
```diff
@@ -26,6 +26,9 @@ def clone(self):
         _, args, kwargs = self.deconstruct()
         return self.__class__(*args, **kwargs)
 
+    def supported(self, connection):
```
Comment (charettes) [inline django/db/models/constraints.py]: Totally opened to a better name here.

Comment (pope1ni) [reply django/db/models/constraints.py]: I have a suggestion. We previously added [`PostgresIndex.check_supported()`](https://github.com/django/django/blob/b33bfc383935cd26e19a2cf71d066ac6edd1425f/django/contrib/postgres/indexes.py#L32-L33) to handle checking support for various index types/flags. 

While that currently takes in `schema_editor`, we could change [this line](https://github.com/django/django/blob/b33bfc383935cd26e19a2cf71d066ac6edd1425f/django/contrib/postgres/indexes.py#L22) to pass `schema_editor.connection` and then we would have some consistency if you also adopted `def check_supported(self, connection):` here.

Comment (pope1ni) [reply django/db/models/constraints.py]: Having said that, `PostgresIndex.check_supported()` doesn't return a boolean, it expects `NotSupportedError` to be raised as it allows checking for multiple conditions. We also don't handle those exceptions intentionally.

I'm not sure that is a deal breaker to making this look similar, however.

**Comment by charettes** [comment]
Something we could do here instead is of special casing `Tags.database` when no tags is specified is too make the check a noop when no `databases` option is specified.

That allows the test runner to only enable checks on databases that were created and the migrate command to only run checks on the specified `--database`. I pushed a WIP that does just that but I'll try to revisit the commit splitting in the next days.

**Comment by charettes** [comment]
I still think it's worth exploring plugging `_check_constraints` and a potential `_check_indices` support on `check_supported` but https://github.com/django/django/pull/12396 is likely less invasive until then.

**Pull request closed without merging**